src/test/java/com/microsoft/azure/relay/SendReceiveTest.java

```diff
@@ -304,38 +296,50 @@ private static void httpRequestSender(String method, String msgExpected, String
 		out.flush();
 		out.close();
 
-		String inputLine;
-		StringBuilder builder = new StringBuilder();
-		BufferedReader inStream = new BufferedReader(new InputStreamReader(conn.getInputStream()));
-		
+		String received = readStringFromStream(conn.getInputStream());
 		assertEquals("Http connection sender did not receive the expected response code.", STATUS_CODE, conn.getResponseCode());
 		assertEquals("Http connection sender did not receive the expected response description.", STATUS_DESCRIPTION, conn.getResponseMessage());
-		
-		while ((inputLine = inStream.readLine()) != null) {
-			builder.append(inputLine);
-		}
-		inStream.close();
 		assertEquals("Http connection sender did not receive the expected response message size.", 
 				msgExpected.length(), 
-				builder.toString().length());
-		assertEquals("Http connection sender did not receive the expected response message.", msgExpected, builder.toString());
+				received.length());
+		assertEquals("Http connection sender did not receive the expected response message.", msgExpected, received);
 	}
 	
-	private static void httpRequestHandlerMultipleSend(RelayedHttpListenerContext context, String firstMsg, String secondMsg, boolean hasPause) {
+	private static void sendResponseMessages(RelayedHttpListenerContext context, String[] messages, boolean hasPause) {
 		RelayedHttpListenerResponse response = context.getResponse();
 		response.setStatusCode(STATUS_CODE);
 		response.setStatusDescription(STATUS_DESCRIPTION);
 		
+		CompletableFuture<?>[] cfs = new CompletableFuture<?>[messages.length];
 		try {
-			response.getOutputStream().write(firstMsg.getBytes());
-			if (hasPause) {
-				Thread.sleep(2500);
+			for (int i = 0; i < messages.length; i++) {
+				cfs[i] = response.getOutputStream().writeAsync(messages[i].getBytes(), 0, messages[i].length());
```

Comment:
>writeAsync [](start = 40, length = 10)

So this calls writeAsync 'i' times but doesn't wait for the previous writeAsync to finish? IMHO that's asking for undefined behavior and customers should't be doing this.

Reply:
Because we have locks on writeAsync, in reality they won't run into the simultaneous write exception thrown from ClientWebSocket. Do you think we should explicitly throw when user writes without waiting then?

---
In reply to: [293086335](https://github.com/Azure/azure-relay-java/pull/46#discussion_r293086335) [](ancestors = 293086335)